crates/sats/src/algebraic_value.rs

```diff
@@ -210,61 +210,46 @@ impl AlgebraicValue {
         Self::Map(Box::new(map))
     }
 
-    /// Returns the [`AlgebraicType`] of the sum value `x`.
-    pub(crate) fn type_of_sum(x: &SumValue) -> AlgebraicType {
-        // TODO(centril, #104): This is unsound!
-        //
-        //   The type of a sum value must be a sum type and *not* a product type.
-        //   Suppose `x.tag` is for the variant `VarName(VarType)`.
-        //   Then `VarType` is *not* the same type as `{ VarName(VarType) | r }`
-        //   where `r` represents a polymorphic variants compontent.
-        //
-        //   To assign this a correct type we either have to store the type with the value
-        //   or alternatively, we must have polymorphic variants (see row polymorphism)
-        //   *and* derive the correct variant name.
-        AlgebraicType::product([x.value.type_of()])
-    }
-
     /// Returns the [`AlgebraicType`] of the product value `x`.
-    pub(crate) fn type_of_product(x: &ProductValue) -> AlgebraicType {
-        AlgebraicType::product(x.elements.iter().map(|x| x.type_of().into()).collect::<Box<[_]>>())
+    pub(crate) fn type_of_product(x: &ProductValue) -> Option<AlgebraicType> {
+        let mut elems = Vec::with_capacity(x.elements.len());
+        for elem in &*x.elements {
+            elems.push(elem.type_of()?.into());
+        }
+        Some(AlgebraicType::product(elems.into_boxed_slice()))
     }
 
     /// Returns the [`AlgebraicType`] of the map with key type `k` and value type `v`.
-    pub(crate) fn type_of_map(val: &MapValue) -> AlgebraicType {
-        AlgebraicType::product(if let Some((k, v)) = val.first_key_value() {
-            [k.type_of(), v.type_of()]
-        } else {
-            // TODO(centril): What is the motivation for this?
-            //   I think this requires a soundness argument.
-            //   I could see that it is OK with the argument that this is an empty map
-            //   under the requirement that we cannot insert elements into the map.
-            [AlgebraicType::never(), AlgebraicType::never()]
-        })
+    pub(crate) fn type_of_map(val: &MapValue) -> Option<AlgebraicType> {
+        let (k, v) = val.first_key_value().and_then(|(k, v)| k.type_of().zip(v.type_of()))?;
+        Some(AlgebraicType::product([k, v]))
     }
 
     /// Infer the [`AlgebraicType`] of an [`AlgebraicValue`].
-    pub fn type_of(&self) -> AlgebraicType {
-        // TODO: What are the types of empty arrays/maps/sums?
+    ///
+    /// This function is partial
+    /// as type inference is not possible for `AlgebraicValue` in the case of sums.
+    /// Thus the method only answers for the decidable subset.
```

Comment:
Why is not possible?

Reply:
Because if you see `VariantZ(data)`, you can't, in general, know what other variants are possible.

Reply:
Oh right, we don't tag a unique id for each `sum`.

Reply:
I think this explanation should be in the comment